Isolated Clusters (#178)
metal-stack/docs: docs/src/overview/isolated-kubernetes.md

## Network Design

In order to be able to restrict ingress and egress internet traffic, but still make it possible to create a working kubernetes cluster we implemented the following network design.

- All strictly required container images are mirrored to a registry which is only accessible from the kubernetes clusters.
- DNS and NTP Servers are produced alongside the registry.
- The `containerd` configuration on every worker node is configured to pull all of the strictly required container images from this private registry mirror.
- DNS and NTP configuration is also adopted to use the DNS and NTP Servers on this private environment.
- A list of networks which are allowed to reach is managed, this list reflects the networks of the cloud provider and is not modifiable by the cluster user. This list usually contains the internet prefixes of the provider and one or more RFC address ranges.

![Network Design](isolated-kubernetes.drawio.svg)

> It is required to attach a additional network to the kubernetes cluster in order to be able to pull container images for the application workloads from private registries.

Reviewer: But the network does not necessarily need to be private, the MPLS networks for instance are not private in our infrastructure.

### Strictly required Container Images

In general the creation of a kubernetes cluster requires the ability to pull container images for several applications which are necessary to make a machine a kubernetes worker node. To mention the most important:

- Kubelet: the main controller on each worker node to manage the workload
- CNI (Container Network Interface): controller and daemon set to setup and run the container networking
- CSI (Container Storage Interface): controller and daemon set to setup and run the container storage
- CoreDNS: DNS for containers
- MetalLB: Service Type Loadbalancer Implementation
- Node-Exporter and Metrics-Exporter: Monitoring for the worker node
- Metal-Stack Addons: for firewall and auditing events

> Because we configured `containerd` as such, that all images which are required, are pulled from the private registry. Because this private registry only contains the strictly required container images, it is not possible anymore to pull container images from public registries.

Author: Thanks, took your suggestion

allow-to-apiserver   deployed
allow-to-dns         deployed
allow-to-ntp         deployed
allow-to-registry    deployed
allow-to-storage     deployed
allow-to-vpn         deployed
```

The purposes of this CWNPs are:

| Rule Name          | Isolation            | Destination                                          | Purpose                                                                                      |
|--------------------|----------------------|------------------------------------------------------|----------------------------------------------------------------------------------------------|
| allow-to-http      | baseline             | 0.0.0.0/0                                            | egress via http                                                                              |
| allow-to-https     | baseline             | 0.0.0.0/0                                            | egress via https                                                                             |
| allow-to-apiserver | restricted, forbidden | IP of the Kubernetes API Server on the control plane | API Server communication of kubelet and other controllers                                    |

> | allow-to-dns       | all                  | IP of the private DNS Server                         | DNS resolution from the Kubernetes worker nodes and containers                               |

Reviewer 2: This sounds like DNS and NTP, even for baseline clusters, is restricted to the private DNS and NTP in the infrastructure cluster. Is that actually the case?

Author: I have to change this table because confusing

| allow-to-storage   | all                  | network of the container storage                     | persistent volumes with the cni driver                                                       |
| allow-to-vpn       | all                  | IP of the vpn endpoint on the control plane          | allow communication from the api server to the kubelet for container logs and container exec |

All of these CWNPs are managed by the gardener-extension-provider-metal, every manual modification will be reverted immediately.

### Internet Access Forbidden

This configuration can only be achieved by creating a new kubernetes cluster, it is not possible to modify a existing cluster (with internet access `baseline` or `restricted`) to this configuration. It is also required to specify the most recent version of kubernetes, older versions of kubernetes are not supported.

Every network access modifications triggered by a cluster user, either by adding or modifying CWNPs or adding a Service Type Loadbalancer, is validated against the list of allowed Networks. If the cluster was created with a additional internal network, this network is part of the allowed networks list.

> 

Reviewer: We configure `containerd` such that all images which are required are pulled from the private registry. Because this private registry only contains the strictly required container images, it is not possible anymore to pull container images from public registries.

With internet access `forbidden` the following restrictions apply:

#### Egress traffic

> Egress traffic is only allowed to the private registry and the DNS and NTP servers. Additional CWNPs can be added to reach destinations in the internal networks if specified.

Reviewer 3: Is it then possible to access an S3 endpoint

Author: Yes

> kubectl get events
8s          Warning   AllocationFailed         service/example-service   Failed to allocate IP for "default/example-service": no available IPs
3s          Warning   SyncLoadBalancerFailed   service/example-service   Error syncing load balancer: failed to ensure load balancer: no default network for ip acquisition specified, acquire an ip for your cluster's project and specify it directly in "spec.loadBalancerIP"
```

### Internet Access Restricted

This configuration can only be achieved by creating a new kubernetes cluster, it is not possible to modify a existing cluster (with internet access `baseline` or `forbidden`) to this configuration. It is also required to specify the most recent version of kubernetes, older versions of kubernetes are not supported.

The same default CWNPs are deployed and the container images are pulled from the private registry. Also DNS and NTP are configured to use the private DNS and NTP servers.
The only difference to the `forbidden` mode is that CWNPs and Service Type Loadbalancers can be created without the restriction that only allowed networks are allowed.

> Pulling container images is theoretically possible, if a cluster creates a CWNP which allows network access to a external registry. But most container registries serve the container images from large CDN Networks which have a lot of ip addresses. Simply adding the ip address of docker.io is therefore not sufficient.

Reviewer: ```suggestion
Pulling container images is theoretically possible if a cluster user creates a CWNP which allows network access to an external registry. But most container registries serve the container images from large CDN networks which have a lot of ip addresses. Simply adding the ip address of docker.io is therefore not sufficient.
```

## Application Container Images

> In order to deploy application containers into a cluster with Internet Access `forbidden` a private registry must be provided. This private registry must be located in list of allowed networks.

Reviewer: ```suggestion
In order to deploy application containers into a cluster with Internet Access `forbidden` a private registry must be provided. This private registry must be located in the list of allowed networks.
```

> The DNS name of the registry must resolve in the public DNS Servers. The registry must be secured with a TLS certificate which is also valid with the ca-certificates from the worker node, e.g. vanilla debian ca-certificates.

Reviewer: ```suggestion
The DNS name of the registry must resolve in the public DNS servers. The registry must be secured with a TLS certificate which is also valid with the ca-certificates from the worker node, e.g. vanilla debian ca-certificates.
```

## Implementation

> To achieve this functionality modifications has been implemented in various components in metal-stack, this includes:

Reviewer: ```suggestion
To achieve this functionality modifications have been implemented in various components in metal-stack, this includes:
```

    // Therefore registries, dns and ntp destinations must be specified in the cloud-profile accordingly-
    // If this is not the case, restricting the access must not be possible.
    // Image overrides for all images which are required to create such a shoot, must be specified. No other images are provided in the given registry.
    // customers can define own rules to access external networks as in the baseline.
    // Service type loadbalancers are also not restricted.
    NetworkAccessRestricted = NetworkAccessType("restricted")
    // NetworkAccessForbidden in this configuration a customer can no longer create rules to access external networks.
    // which are outside of a given list of allowed networks. This is enforced by the firewall.
    // Service type loadbalancers are also not possible to open a service ip which is not in the list of allowed networks.
    // This is also enforced by the firewall.
    NetworkAccessForbidden = NetworkAccessType("forbidden")
)
```

A sample Shoot Spec:

```yaml
---
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: isolated
  namespace: sample
spec:
  provider:
    type: metal
      controlPlaneConfig:

>         networkAccessType: forbidden
> ```

Reviewer: ```suggestion
        networkAccessType: forbidden
...
```
```

    // DNSServers
    DNSServers []string
    // NTPServers
    NTPServers []string
    // The registry which serves the images required to create a shoot.
    RegistryMirrors []RegistryMirror
}

// AllowedNetworks is a list of networks which are allowed to connect in restricted or forbidden NetworkIsolated clusters.
type AllowedNetworks struct {
    // Ingress defines a list of networks which are allowed for incoming traffic like service type loadbalancer
    // to allow all you must specify 0.0.0.0/0 or ::/0
    Ingress []string
    // Egress defines a list of networks which are allowed for outgoing traffic
    // to allow all you must specify 0.0.0.0/0 or ::/0
    Egress []string
}

type RegistryMirror struct {
    // Name describes this server
    Name string
    // Endpoint is typically the url of the registry in the form https://hostname
    Endpoint string
    // IP is the ipv4 or ipv6 address of this server
    IP string
    // Port at which port the service is reachable
    Port int32
    // This Registry Mirror mirrors the following registries
    MirrorOf []string   
}
```

A sample configuration in the CloudProfile would look like:

```yaml

>     network-isolation:

Reviewer: ```suggestion
network-isolation:
```

    allowedNetworks:
        egress:
        - 1.2.3.0/24 # Internet CIDR of the Provider
        - 100.64.0.0/10
        - 10.0.0.0/8
        ingress:
        - 100.64.0.0/10
    dnsServers:
        - "1.2.3.1"
        - "1.2.3.2"
        - "1.2.3.3"
    ntpServers:
        - "1.2.3.1"
        - "1.2.3.2"
        - "1.2.3.3"
    registryMirrors:

>         - name: test registry

Reviewer: ```suggestion
      - name: test registry
```

If a cluster is either configured with `restricted` or `forbidden`, the configuration of containerd will be created as such:

config.toml

```toml
imports = ["/etc/containerd/conf.d/*.toml"]
version = 2

[plugins."io.containerd.grpc.v1.cri".registry]
  [plugins."io.containerd.grpc.v1.cri".reg
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
      endpoint = ["https://some.private.registry"]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."quay.io"]
      endpoint = ["https://some.private.registry"]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.lightbitslabs.com"]
      endpoint = ["https://some.private.registry"]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."eu.gcr.io"]
      endpoint = ["https://some.private.registry"]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."ghcr.io"]
      endpoint = ["https://some.private.registry"]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."registry.k8s.io"]
      endpoint = ["https://some.private.registry"]
    [plugins."io.containerd.grpc.v1.cri".registry.mirrors."r.metal-stack.io"]
      endpoint = ["https://some.private.registry"]
```

> DNS and NTP must also be adopted according the configuration in the CloudProfile.

Reviewer: ```suggestion
DNS and NTP must also be adopted according to the configuration in the CloudProfile.
```

### Firewall Controller Manager and Firewall Controller

> The  Firewall Controller Manager changed the FirewallSpec to configure the Firewall Controller which must enforce the restrictions regarding allowed networks.

Reviewer: ```suggestion
The Firewall Controller Manager has extended the FirewallSpec to configure the Firewall Controller which must enforce the restrictions regarding allowed networks.
```

    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`   

    Spec   PolicySpec   `json:"spec,omitempty"`
    Status PolicyStatus `json:"status,omitempty"`
}

// PolicyDeploymentState describes the state of a CWNP deployment
type PolicyDeploymentState string

const (
    // PolicyDeploymentStateDeployed the CWNP was deployed to a native nftable rule
    PolicyDeploymentStateDeployed = PolicyDeploymentState("deployed")
    // PolicyDeploymentStateIgnored the CWNP was not deployed to a native nftable rule because it is outside of allowed networks
    PolicyDeploymentStateIgnored  = PolicyDeploymentState("ignored")
)

// PolicyStatus defines the observed state for CWNP resource
type PolicyStatus struct {
    // FQDNState stores mapping from FQDN rules to nftables sets used for a firewall rule.
    // Key is either MatchName or MatchPattern
    // +optional
    FQDNState FQDNState `json:"fqdn_state,omitempty"`   
    // State of the CWNP, can be either deployed or ignored
    State PolicyDeploymentState `json:"state"`  
    // Message describe why the state changed
    Message string `json:"message,omitempty"`
}
```

### Cloud Controller Manager

> This component was adopted to allow to be started without a default network specified. This was actually always the internet network and if no ip address was specified in the Service Type Loadbalancer, one ip was allocated from this default network. For restricted and isolated clusters this is not provided and a cluster user must always specify this ip to get a working loadbalancer.

Reviewer: ```suggestion
This component was adopted to allow to be started without a default network specified. This was actually always the internet network and if no ip address was specified in the Service Type Loadbalancer, one ip was allocated from this default network. For isolated clusters this is not provided and a cluster user must always specify this ip to get a working loadbalancer.
```

### OCI Mirror

> The [OCI Mirror](https://github.com/metal-stack/oci-mirror) is a new application which acts as a scheduled job which pulls a given list of container images and pushes them to a private registry. The detailed description can be read on the project website.

Reviewer: ```suggestion
The [OCI Mirror](https://github.com/metal-stack/oci-mirror) is a new application which acts as a scheduled job which pulls a given list of container images and pushes them to a private registry (which will then serve as the private registry mirror). The detailed description can be read on the project website.
```
